### DISCUSSION TITLE
Hook for export functionality in App layout

### DISCUSSION THREAD

**Discussion by christianrr** [directus/directus#6656] [Feature Requests]
I was wondering if it would be possible to use the extension SDK to provide an extension method for the export option within App layouts. Ideally, we could define two parts there: Firstly, a configuration part that defines the available export options (e.g. CSV only), and secondly, some custom endpoint that is executed whenever the export button in the layout is triggered (user choices like export type, etc. as parameter).

This would allow us to create fully customized exports, including relational data. We often have the issue, that end customers would expect only a subset of data for export or that relations are included, since endusers usually don't care about UUIDs. 

We usually create a custom endpoint for export then, which is delivered to the enduser by using e.g. a button interface or custom page. However, since export already has already its dedicated space in UI, it would be nice to utilize this spot and customize the functionality.

**Comment by rijkvanzanten** [comment]
Also sounds like we could improve the UI for the export functionality to begin with. You can already export relational data through the regular export functionality f.e. 🙂

**Reply by benhaynes** [reply]
Would it make sense to allow registering new custom events for hooks?

**Reply by christianrr** [reply]
> Also sounds like we could improve the UI for the export functionality to begin with. You can already export relational data through the regular export functionality f.e. 🙂

@rijkvanzanten Oh, did not know that. How should that work exactly? We can only see primary keys in exports of tables with relational fields.

**Reply by christianrr** [reply]
> Would it make sense to allow registering new custom events for hooks?

I figured out that the export in the app triggers the items endpoint with the export option. For our scenarios it would be perfectly fine, if there would be a hook to intercept the default export behavior and replace it with some custom code.

What we would need as context parameters in this hook are the UI settings (export type, filters, filter checkbox) as well as the collection name.

**Reply by rijkvanzanten** [reply]
@christianrr The `export` parameter effectively saves whatever the normal API response is to a file. You can add any combination of `fields`, `filters`, etc to control the output 👍🏻

**Reply by christianrr** [reply]
@rijkvanzanten Got that. So you meant that we could improve the UI for export with some field options or similar?

**Reply by rijkvanzanten** [reply]
Yeah! That's what I'm thinking :) Currently, the export sidebar detail relies on the layout to tell it what to export, so the fact that you can only export the primary key of a related record comes solely from the fact that the tabular layout currently doesn't allow for selecting "deep" columns from the related item 👍🏻

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
